internal/provider/resource_ko_build.go

```diff
@@ -18,105 +17,239 @@ import (
 	"github.com/google/ko/pkg/build"
 	"github.com/google/ko/pkg/commands/options"
 	"github.com/google/ko/pkg/publish"
-	"github.com/hashicorp/go-cty/cty"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/listdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/listplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
 )
 
-const (
-	defaultBaseImage = "cgr.dev/chainguard/static"
-	version          = "devel"
-)
+var _ resource.Resource = &BuildResource{}
+var _ resource.ResourceWithImportState = &BuildResource{}
 
-var validTypes = map[string]struct{}{
-	"spdx":         {},
-	"cyclonedx":    {},
-	"go.version-m": {},
-	"none":         {},
+func NewBuildResource() resource.Resource {
+	return &BuildResource{}
+}
+
+// BuildResource defines the resource implementation.
+type BuildResource struct {
+	popts Opts
+}
+
+// BuildResourceModel describes the resource data model.
+type BuildResourceModel struct {
+	// Inputs
+	Importpath types.String `tfsdk:"importpath"`
+	WorkingDir types.String `tfsdk:"working_dir"`
+	Platforms  types.List   `tfsdk:"platforms"`
+	BaseImage  types.String `tfsdk:"base_image"`
+	SBOM       types.String `tfsdk:"sbom"`
+	Repo       types.String `tfsdk:"repo"`
+
+	// Computed attributes
+	Id       types.String `tfsdk:"id"`
+	ImageRef types.String `tfsdk:"image_ref"`
+
+	// Overridden by provider opts
+	repo     string
+	bare     bool
+	keychain authn.Keychain
+	version  string
+}
+
+func (r *BuildResourceModel) update(popts Opts) {
+	r.version = popts.version
+	if !r.Repo.IsNull() { // ko_build.repo being set takes precedence over provider opts, and makes it bare
+		r.repo = r.Repo.ValueString()
+		r.bare = true
+	} else if popts.repo != "" { // provider opts takes precedent over env var
+		r.repo = popts.repo
+	} else { // env var is last resort
+		r.repo = os.Getenv("KO_DOCKER_REPO")
```

Comment:
I feel like with the wisdom we've gained (and hindsight) that we should probably deprecate support for this.  It feels like an anti-pattern for IaC.

I don't think we need to do anything in this change, but maybe we should start to warn folks when we're using this that they should use an explicit repo on the provider or rule (possibly set via an env var like `TF_VAR_BLAH`)

Reply:
I agree. https://github.com/ko-build/terraform-provider-ko/issues/70